Add documentation for public.postscript.hints

readroberts opened the pull request:
First pass documentation, do not hesitate to edit. I propose changing the rules for creating the hint dict "id" value by dropping the "base:<component glyph name"  item when parsing a component glyph.

readroberts committed:
Add documentation for public.postscript.hints

schriftgestalt commented on versions/ufo3/lib.plist.md:
```diff
@@ -29,6 +29,49 @@ This defines a preferred glyph name to Postscript glyph name mapping for glyphs
+
+| key         | value type | description                                                                                                                                                                                                                                                  |
+|-------------|------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| id          | string     | Hash of glyph outlines. This is computed when the glyph is hinted. It is used to determine if the glyph outline has been changed since the glyph was hinted: if it has, then the hint dict for the glyph should be deleted. See "Hint ID Computation" below. |
```
So if one adjusts the spacing a bit, all hints are to be deleted? Isn't there a more robust hashing algorithm. Something that encodes the proportion/relations of the nodes instead of the absolute coordinate. One first step could be to store the relative position to the previous point. That would at least take care of the spacing.

readroberts replied:
Although the hint substitution action is tied to specific points, stem positions are not: they are only relative to each other, and to the glyph origin at (0,0). Hence, shifting the outlines by changing the LSB does mean that all the hints are invalidated. 

However, changing the advance width alone does not invalidate the hints. I would further modify the proposal by omitting the width from the start of the hash string. 

I don't see any other advantage to computing relative coordinate values before computing the hash: this requires more processing time, as the values must be converted from string to real numbers, and the relative positions calculated, while the only benefit is making the pre-hash string somewhat more compact.

schriftgestalt commented on versions/ufo3/lib.plist.md:
```diff
@@ -29,6 +29,49 @@ This defines a preferred glyph name to Postscript glyph name mapping for glyphs
+#### public.postscript.hints
+
+This key provides a mapping from glyph names to sets of PostScript hints for the
+default glyph layer. The mapping is stored as a dictionary with glyph names as
```
Why not store the same data in the lib in the glif? That would make it much simpler and removes the possibility to loose track when the glyph names change.

readroberts replied:
That makes sense to me. I had thought Tal was requesting that this be in lib.plist file, but I see that is not the case. I will modify the proposal to store the hint key in the GLIF file lib.

typesupply commented:
@readroberts I agree with @schriftgestalt on moving this to the individual GLIF lib. As for the hash, I'll defer to your expertise.

One question: do the `pointTag` strings correspond to a point's `name` attribute or `identifier` attribute? Referencing the `name` attribute makes sense as that is backwards compatible to UFO 1. We designed the `identifier` attribute in GLIF 2/UFO 3 specifically for references like you are using here. It's similar to the `name` attribute but isn't to be shown to the user and is supposed to be persistent once assigned. Perhaps the next version of this spec could make use of the `identifier` attribute. That brings me to a suggestion...

I think it would be a good addition to add a `formatVersion` key, with a value of `1` to the `Hint Dict` object. If the key isn't defined, the default value is `1`. This will give us a clear way to distinguish versions of this data format.

Once the hint dict is moved to the individual GLIF, I'm fine with merging this into the spec.

benkiel commented:
As things using UFO 1 are few (any?) and far between, perhaps it's better to use ``identifier``?

typesupply commented:
> As things using UFO 1 are few (any?) and far between

Oh, I meant that `identifier` isn't in UFO 1 or 2 and there are a lot of UFO 2 files out there. Sorry for the mixup.

benkiel commented:
Dur, right. Nevermind, ``name`` is good and +1 for a formatVersion so that it can transition to using ``identifier`` for version 2 for UFO3

schriftgestalt commented:
Can we also agree on a bit more high level structure. In most cases we don’t need hint replacement but have the hints attached to points like the Links in FontLab. That would be more stable, specially during design when the outlines change a lot. 

It could look something like this:
```XML
	<key>linkList</key>
	<array>
		<dict>
			<key>type</key>
			<string>hstem</string>
			<key>origin</key>
			<string>nodeName1</string>
			<key>target</key>
			<string>nodeName2</string>
		</dict>
		<dict>
			<key>type</key>
			<string>vstem</string>
			<key>origin</key>
			<string>nodeName1</string>
			<key>target</key>
			<string>nodeName4</string>
		</dict>
	</array>
```

readroberts commented:
I will update my PR to move the hint dict to the GLIF lib, add the formatVersion key, and to be clear the the pointTag is supposed to match the point name attribute. 

I am reluctant to tie hint substitution to specific points. One of the features of CFF/Type hinting is that a stem hint is not tied to a point, but instead specifies a vertical or horizontal zone, with some fuzz on the edges, so that the hint zone can capture several points, and on different contours. If a hint zone gets moved by moving  a point, the hint zone may become invalid for other points that it used to capture. This why it is a good idea to re-hint whenever any editing changes are made.

readroberts committed:
Revert "Add documentation for public.postscript.hints"

This reverts commit 17773b2ba38a59465f222037f95617e7111bd00a.

readroberts committed:
Postscript hints. Move the definition of the PostScript hints to the glif file 'lib' dict as a public key. Add the "formatVersion" key. Note that the hint set pointTag must match a point name attribute. Add a little more info about use of the glyph outline hash string. Remove the glyph width from the source string for the hash. Remove the "base:" prefix for a component hash from the source string for the hash.

moyogo commented:
Can this be merged or do we still need some changes for consensus?

moyogo commented on versions/ufo3/glyphs/glif.md:
```diff
@@ -351,6 +386,50 @@ This key is used for representing the "vertical origin" 'y' coordinate used for
       <string>1,0,0,0.5</string>
+      <key>public.postscript.hints</key>
+    <dict>
+        <key>formatVersion<key><string>1</string>
```
The key element `<key>formatVersion<key>` isn’t closed properly.

typesupply commented:
I'm ready to merge whenever @readroberts is.

moyogo commented:
It might be helpful to have a more detailed description of the Hint ID computation. I couldn’t reproduce the Hint ID of the example glyph.

There would also be hash collision in some extreme corner cases with the Hind ID computation as I understand it.
When whitespace is omitted some pairs of x and y values may be give the same has, for example "1.234.5" may be 1.2 and 34.5 in one outline, but 1.23 and 4.5 in another.
It may be fine to overlook these cases.

readroberts commented:
@moyogo Can you post the Hint ID string that you derived from the example glyph? My test code reproduces the Hint ID string shown in the example. Since the string is less than 256 chars, it is not hashed, and it should be easy to see to see the differences. I'd like to make sure we resolve any ambiguity in the Hint ID computation.

readroberts commented:
I would also like to propose change in the Hint ID computation. The current AFDKO logic includes the glyph width in the Hint ID string. I dropped this from the public spec for hint ID computation because it does not affect the hinting of a glyph, but I would now like to add it back. If the hinted glyph is the primary source glyph for all operations, then omitting the width makes sense. However, in the AFDKO implementation, and I suspect others as well, the hinted glyph file is stored in a separate layer from the default layer where glyph edits happen. A glyph in this 'processed' layer must be invalidated when the width in the original source layer is changed. This is not necessary, as code that knows to prefer the 'processed' layer can also know to add the width to the hint ID string before comparing to the source glyph, but it is convenient to not have to do this. Comments?

benkiel commented:
@readroberts sounds good

moyogo commented:
@readroberts the Hint ID in the example is "64bf4987f05ced2a50195f971cd924984047eb1d79c8c43e6a0054f59cc85dea23a49deb20946a4ea84840534363f7a13cca31a81b1e7e33c832185173369086".

Is it supposed to be "237152193187curve1341877418730150curve3088302374-10curve134-10193-1023725curve23788" with "offcurve" left out or "offcurve237152offcurve193187curve134187offcurve74187offcurve30150curve3088offcurve3023offcurve74-10curve134-10offcurve193-10offcurve23725curve23788" with "offcurve"?

Also, I guess you meant the Hint ID string being less that 128, as in the pull request, instead 256 characters in your comment.

Would it make sense to require the layer used for the ps hinting to have a width defined instead of getting it from another layer?

readroberts committed:
[public.postscript.hints]. Updates to address need for width in glyph ID tag, and  comments in this thread.

-Added 'width' value to start of ID key, in order to invalidate the glyph if the source glyph width is changed.
- Added comma between point values, and whitespace before offcurve  point values. By delimiting values, we do reduce the change of hash collisions, and certainly make the initial ID string easier to read.
- Fixed ID string to match what is actually produced for this glyph data.
- fixed hint values to match output form the current afdko autohint tool for this glyph data, and the BlueValues from Source-SerifPro-Regular.

readroberts committed:
Revert "[public.postscript.hints]. Updates to address need for width in glyph ID tag, and  comments in this thread."

This reverts commit 07aa23450f0331ddb7018e7eae065d155ff62c2f.

readroberts committed:
[public.postscript.hints]. Updates to address need for width in glyph ID tag, and comments in this thread.

Try #2. I forgot that that the 'Texts' program can make many changes to an 'md' file, beyond the small typo I wanted to fix.

-Added 'width' value to start of ID key, in order to invalidate the glyph if the source glyph width is changed.
- Added comma between point values, and whitespace before offcurve point values. By delimiting values, we do reduce the change of hash collisions, and certainly make the initial ID string easier to read.
- Fixed ID string to match what is actually produced for this glyph data.
- fixed hint values to match output form the current afdko autohint tool for this glyph data, and the BlueValues from Source-SerifPro-Regular.

readroberts commented:
I am ready to see this merged, unless there are more comments on the last set of changes.

typesupply commented:
I'll wait a couple of days for feedback and then merge. Thanks, Read!

moyogo commented:
Sorry for the late reply. It looks good to me.
